sacc/covariance.py

```diff
@@ -0,0 +1,466 @@
+        return subclass.from_hdu(hdu)
+
+    @classmethod
+    def make(cls, cov):
+        """Make an appropriate covariance object from the matrix info itself.
+
+        You can pass in a list of covariance blocks for a block-diagonal, covariance
+        a 1D array for a diagonal covariance, or a full matrix.
+
+        A different subclass is returned for each of these cases.
+
+        Parameters
+        ----------
+        cov: list[array] or array
+            If a list, the total length of all the arrays in it
+            should equal n.  If an array, it should be either 1D of
+            length n or 2D of shape (n x n).
+
+        n: int
+            length of the data vector to which this covariance applies
+        """
+        if isinstance(cov, list):
+            s = 0
+            for block in cov:
+                block = np.atleast_2d(block)
+                if (block.ndim != 2) or (block.shape[0] != block.shape[1]):
+                    raise ValueError(f"Covariance block has wrong size or shape {block.shape}")
+                s += block.shape[0]
+            return BlockDiagonalCovariance(cov)
+        else:
+            cov = np.array(cov).squeeze()
+            if cov.ndim == 1:
+                return DiagonalCovariance(cov)
+            if (cov.ndim != 2) or (cov.shape[0] != cov.shape[1]):
+                raise ValueError(f"Covariance is not a 2D square matrix - shape: {cov.shape}")
+            return FullCovariance(cov)
+
+
+class FullCovariance(BaseCovariance, cov_type='full'):
+    """
+    A covariance subclass representing a full matrix with correlations
+    anywhere.  Represented as an n x n matrix.
+
+    Attributes
+    ----------
+
+    size: int
+        the length of the corresponding data vector
+
+    covmat: 2D array
+        The matrix itself, of shape (size x size)
+    """
+    def __init__(self, covmat):
+        self.covmat = np.atleast_2d(covmat)
+        self.size = self.covmat.shape[0]
+
+    def to_hdu(self):
+        """
+        Make an astropy FITS HDU object with this covariance in it.
+        This is represented as an image.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        hdu: astropy.fits.ImageHDU instance
+            HDU that can be used to reconstruct the object.
+        """
+        hdu = fits.ImageHDU(self.covmat)
+        hdu.header['EXTNAME'] = 'covariance'
+        hdu.header['SACCTYPE'] = 'cov'
+        hdu.header['SACCCLSS'] = self.cov_type
+        hdu.header['SIZE'] = self.size
+        return hdu
+
+    @classmethod
+    def from_hdu(cls, hdu):
+        """
+        Load a covariance object from the data in the HDU
+
+        Parameters
+        ----------
+        hdu: astropy.fits.ImageHDU instance
+
+        Returns
+        -------
+        cov: FullCovariance
+            Loaded covariance object
+        """
+        C = hdu.data
+        return cls(C)
+
+
+    def masked(self, mask):
```

Comment:
We should change this verb too.

Reply:
How about `new_cov = cov.keeping(indices)`?

Reply:
or `keep_indices` like the other one?

Reply:
`keep_indices` is in-place (and so imperative), so it makes sense to use a different verb form (tense? mood?) for this, which returns a new instance.

Reply:
Maybe `keeping_indices`

Reply:
Ahhhh ok. No reason to obsess over this `with_only_indices` would do the trick or something similar.

Reply:
yup. `keeping_indices` is fine.

Reply:
Changed.  And test added.